infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java

```diff
@@ -0,0 +1,218 @@
+            .build();
+
+    @Getter
+    public final StreamLog streamLog;
+
+    @Getter
+    private final AddressSpaceView addressSpaceView;
+
+    @Override
+    public CompletableFuture<TransferBatchResponse> transfer(TransferBatchRequest transferBatchRequest) {
+        return readRecords(transferBatchRequest, 0)
+                .thenApply(records -> writeRecords(records, streamLog, maxOverwriteExceptions))
+                .exceptionally(error -> TransferBatchResponse
+                        .builder()
+                        .status(FAILED)
+                        .causeOfFailure(Optional.of(new StateTransferBatchProcessorException(error)))
+                        .build()
+                );
+    }
+
+    /**
+     * Reads data entries by utilizing the replication protocol. If there are errors, retries.
+     *
+     * @param transferBatchRequest The transferBatchRequest of addresses to be read.
+     * @param retries              The number of retries.
+     * @return A result of reading records.
+     */
+    CompletableFuture<ReadBatch> readRecords(TransferBatchRequest transferBatchRequest, int retries) {
+        return CompletableFuture
+                .supplyAsync(() -> {
+                            Supplier<Map<Long, ILogData>> readSupplier = () -> addressSpaceView
+                                    .simpleProtocolRead(transferBatchRequest.getAddresses(), readOptions);
+
+                            return Result
+                                    .of(readSupplier)
+                                    .mapError(StateTransferException::new);
+                        }
+                ).thenApply(readResult ->
+                        readResult.map(records -> checkReadRecords(
+                                transferBatchRequest.getAddresses(),
+                                records,
+                                transferBatchRequest.getDestination()
+                        ))
+                ).thenCompose(checkedReadResult -> {
+                    if (checkedReadResult.isError()) {
+                        return retryReadRecords(transferBatchRequest, retries);
+                    } else {
+                        ReadBatch readBatch = checkedReadResult.get();
+                        if (readBatch.getStatus() == ReadBatch.ReadStatus.FAILED) {
+                            return retryReadRecords(readBatch.createRequest(), retries);
+                        } else {
+                            return CompletableFuture.completedFuture(readBatch);
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Handle read errors. Exponential backoff policy is utilized.
+     *
+     * @param transferBatchRequest TransferBatchRequest of addresses to be read.
+     * @param retriesTried         Configurable number of retries for the current transferBatchRequest.
+     * @return Future of the result of the retry.
+     */
+    CompletableFuture<ReadBatch> retryReadRecords
```

Comment:
You are not invalidating layout when retrying, but if there is a WrongEpochException in read, then retry always fails, right?

Reply:
Yes unlike other runtime exceptions it will fail because we create the transfer segments state list based on the initial layout before running a transfer. If we invalidate a layout, we would have to account for the new changed state midway transfer. This issue is addressed in the next PR.